app/controllers/application_controller.rb

```diff
@@ -133,4 +133,60 @@ def implicit_authorization_target
   def identification_params
     params.permit(:id)
   end
+
+  def upload_criteria(assignment, data)
```

Comment:
I like that you've factored this out of the criteria controller so that we can use it in multiple places. However I don't think that this belongs in the ApplicationController. The ApplicationController is for methods that affect all application routes, this one is specific to criteria. 

I would move this function to a new helper (app/helpers/criteria_helper.rb)

Reply:
The other thing I'd like you to think about is the fact that you now have a transaction in a transaction. If an error is raised during the criteria creation do you want everything to roll back or just the criteria creation step?

Also, this method displays flash messages which is fine but does not follow the pattern we established earlier of raising errors and reporting the errors as flash messages _at the end_.

Instead of just moving this function, think about how you can tweak it a bit so that way you're handling transactions and error messages stays consistent.